File: src/main/java/WaTor.java
import scala.Tuple2;
import scala.Tuple3;

import java.applet.Applet;
import java.awt.*;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.List;
import java.util.stream.Collectors;

class WaTorCanvas extends Canvas {

    public static final int DEFAULT_WIDTH = 40;
    public static final int DEFAULT_HEIGHT = 40;
    public static final int DEFAULT_TIMEOUT = 100;
    public static final int DEFAULT_FISH_COUNT = 50;
    public static final int DEFAULT_SHARK_COUNT = 5;
    public static final int DEFAULT_SHARK_ENERGY = 20;
    public static final int DEFAULT_FISH_REPR_TIME = 8;
    public static final int DEFAULT_SHARK_REPR_TIME = 15;

    public static final Path pathToPhasePortraitCSV = Paths.get("D:\\phasePortrait.csv");
    public static final Path pathToFishPopulationCSV = Paths.get("D:\\fishPopulation.csv");
    public static final Path pathToSharkPopulationCSV = Paths.get("D:\\sharkPopulation.csv");

    private int step = 0;
    private World world;
    private Graphics graphics;
    private Image screenImage;

    private int width;
    private int height;
    private int fishCount;
    private int fishReprTime;
    private int sharkCount;
    private int sharkReprTime;
    private int sharkEnergy;

    private boolean running = false;
    private int timeout = DEFAULT_TIMEOUT;

    private java.util.List<Integer> sharks = new ArrayList<>();
    private java.util.List<Integer> fishes = new ArrayList<>();

    public WaTorCanvas(int width, int height, int fishCount, int fishReprTime, int sharkCount, int sharkReprTime, int sharkEnergy) {
        this.width = width;
        this.height = height;
        this.fishCount = fishCount;
        this.fishReprTime = fishReprTime;
        this.sharkCount = sharkCount;
        this.sharkReprTime = sharkReprTime;
        this.sharkEnergy = sharkEnergy;
        world = new World(new Tuple2<>(height, width), new Tuple2<>(fishCount, fishReprTime), new Tuple3<>(sharkCount, sharkReprTime, sharkEnergy));
        sharks.add(world.countSharks());
        fishes.add(world.countFishes());
    }

    public WaTorCanvas(){
        this(
            DEFAULT_WIDTH,
            DEFAULT_HEIGHT,
            DEFAULT_FISH_COUNT,
            DEFAULT_FISH_REPR_TIME,
            DEFAULT_SHARK_COUNT,
            DEFAULT_SHARK_REPR_TIME,
            DEFAULT_SHARK_ENERGY
        );
    }

    /**
     * re-initialize applet with new parameters
     */
    public void reload(int width, int height, int fishCount, int fishReprTime, int sharkCount, int sharkReprTime, int sharkEnergy){
        fishes.clear();
        sharks.clear();
        this.width = width;
        this.height = height;
        world = new World(new Tuple2<>(height, width), new Tuple2<>(fishCount, fishReprTime), new Tuple3<>(sharkCount, sharkReprTime, sharkEnergy));
    }

    /**
     * re-initialize applet with default parameters
     */
    public void reload(){
        sharks.clear();
        fishes.clear();
        this.width = DEFAULT_WIDTH;
        this.height = DEFAULT_HEIGHT;
        this.timeout = DEFAULT_TIMEOUT;
        world = new World(
                new Tuple2<>(DEFAULT_WIDTH, DEFAULT_HEIGHT),
                new Tuple2<>(DEFAULT_FISH_COUNT, DEFAULT_FISH_REPR_TIME),
                new Tuple3<>(DEFAULT_SHARK_COUNT, DEFAULT_SHARK_REPR_TIME, DEFAULT_SHARK_ENERGY)
        );
    }

    public void setRunning(boolean r) {
        running = r;
        repaint();
    }

    public void update(Graphics g) {
        final Dimension d = getSize();
        if (screenImage == null || screenImage.getWidth(null) != d.width || screenImage.getHeight(null) != d.height) {
            screenImage = createImage(d.width, d.height);
            if (graphics != null) {
                graphics.dispose();
            }
            graphics = screenImage.getGraphics();
        }
        graphics.setColor(getBackground());
        graphics.fillRect(0, 0, d.width, d.height);
        paint(graphics);
        g.drawImage(screenImage, 0, 0, this);
    }

    public void paint(Graphics g) {
        if (!running) {
            paintOcean(g);
            return;
        }
        try {
            Thread.sleep(timeout);
        } catch (InterruptedException ignored){}
        world.forward(step++);
        sharks.add(world.countSharks());
        fishes.add(world.countFishes());
        repaint();
        paintOcean(g);
    }

    /**
     * show all ocean on each moment of time
     * green square - fish
     * red square - shark
     * blue square - empty
     */
    private void paintOcean(Graphics g) {
        // calculate number of pixels in one ocean square
        final Dimension d = getSize();

        final int minDim = Math.min(d.width, d.height);
        final int oceanSquareSize = Math.min(minDim / height, minDim / width);

        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                if (world.ocean()[i][j] == null) {
                    g.setColor(Color.cyan);
                } else if (world.ocean()[i][j].isFish()) {
                    g.setColor(Color.green);
                } else if (world.ocean()[i][j].isShark()) {
                    g.setColor(Color.red);
                }
                g.fillRect(j * oceanSquareSize, i * oceanSquareSize, oceanSquareSize - 1, oceanSquareSize - 1);
            }
        }
    }

    public int getFishCount() {
        return fishCount;
    }

    public int getFishReprTime() {
        return fishReprTime;
    }

    public int getSharkCount() {
        return sharkCount;
    }

    public int getSharkReprTime() {
        return sharkReprTime;
    }

    public int getSharkEnergy() {
        return sharkEnergy;
    }

    public int getTimeout() {
        return timeout;
    }

    public String setTimeout(String timeout) {
        try{
            this.timeout = Integer.valueOf(timeout);
        } catch (Exception e){
            this.timeout = DEFAULT_TIMEOUT;
        }
        return String.valueOf(this.timeout);
    }

    /**
     * saves data from each moment of time into files
     */
    public void saveData() {
        processData(pathToPhasePortraitCSV, toCSVLine(createPhasePortrait()));
        processData(pathToFishPopulationCSV, toCSVLine(createPopulation(fishes)));
        processData(pathToSharkPopulationCSV, toCSVLine(createPopulation(sharks)));
    }

    private void processData(final Path path, final Collection<String> data){
        try {
            Files.deleteIfExists(path);
            Files.createFile(path);
            Files.write(path, data);
        } catch (IOException ignored) {}
    }

    private Collection<String> toCSVLine(final Collection<Tuple2<Integer, Integer>> lines){
        final String line0 = "" + lines.stream().map(t -> t._1().toString()).collect(Collectors.joining(","));
        final String line1 = lines.stream().map(t -> t._2().toString()).collect(Collectors.joining(","));
        return Arrays.asList(line0, line1);
    }

    private List<Tuple2<Integer, Integer>> createPhasePortrait(){
        final List<Tuple2<Integer, Integer>> list = new ArrayList<>();
        for (int i = 0; i < sharks.size(); i++) {
            list.add(new Tuple2<>(sharks.get(i), fishes.get(i)));
        }
        return list;
    }

    private List<Tuple2<Integer, Integer>> createPopulation(final List<Integer> counts){
        final List<Tuple2<Integer, Integer>> list = new ArrayList<>();
        for (int i = 0; i < sharks.size(); i++) {
            list.add(new Tuple2<>(i, counts.get(i)));
        }
        return list;
    }
}

public class WaTor extends Applet {
    private WaTorCanvas wc;
    private TextField timeoutField;
    private TextField initFishField;
    private TextField initSharkField;
    private TextField fishBreedScrollField;
    private TextField sharkBreedScrollField;
    private TextField sharkStarveScrollField;

    /**
     * initialize applet
     */
    public void init() {
        setLayout(new FlowLayout());
        setSize(650, 475);

        wc = new WaTorCanvas();
        wc.setSize(400, 400);
        add(wc);

        final Panel p = new Panel();
        p.setLayout(new GridLayout(11, 1, 3, 3));
        final Button start = new Button("Start");
        start.addActionListener(e -> {
            wc.setRunning(true);
            timeoutField.setText(wc.setTimeout(timeoutField.getText()));
        });
        p.add(start);
        final Button stop = new Button("Stop");
        stop.addActionListener(e -> wc.setRunning(false));
        p.add(stop);
        final Button reset = new Button("Reset");
        reset.addActionListener(e -> {
            wc.setRunning(false);
            wc.reload();
            initFishField.setText(String.valueOf(wc.getFishCount()));
            fishBreedScrollField.setText(String.valueOf(wc.getFishReprTime()));
            initSharkField.setText(String.valueOf(wc.getSharkCount()));
            sharkBreedScrollField.setText(String.valueOf(wc.getSharkReprTime()));
            sharkStarveScrollField.setText(String.valueOf(wc.getSharkEnergy()));
            timeoutField.setText(String.valueOf(wc.getTimeout()));
        });
        p.add(reset);
        final Button applyNewSettings = new Button("Apply new Settings");
        applyNewSettings.addActionListener(e -> {
            wc.setRunning(false);
            wc.reload(
                    WaTorCanvas.DEFAULT_WIDTH,
                    WaTorCanvas.DEFAULT_HEIGHT,
                    new Integer(initFishField.getText()),
                    new Integer(fishBreedScrollField.getText()),
                    new Integer(initSharkField.getText()),
                    new Integer(sharkBreedScrollField.getText()),
                    new Integer(sharkStarveScrollField.getText())
            );
            timeoutField.setText(wc.setTimeout(timeoutField.getText()));
        });
        p.add(applyNewSettings);

        final Button save = new Button("Save Data");
        save.addActionListener(e -> wc.saveData());
        p.add(save);

        final Panel fish = new Panel();
        fish.add(new Label("Initial Fish"));
        initFishField = new TextField(Integer.toString(wc.getFishCount()), 5);
        fish.add(initFishField);
        p.add(fish);

        final Panel shark = new Panel();
        shark.add(new Label("Initial Sharks"));
        initSharkField = new TextField(Integer.toString(wc.getSharkCount()), 5);
        shark.add(initSharkField);
        p.add(shark);

        final Panel timeout = new Panel();
        timeout.add(new Label("Repaint Timeout"));
        timeoutField = new TextField(Integer.toString(wc.getTimeout()), 5);
        timeout.add(timeoutField);
        timeoutField.addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                if (e.getKeyCode() == KeyEvent.VK_ENTER) {
                    timeoutField.setText(wc.setTimeout(timeoutField.getText()));
                }
            }
        });
        p.add(timeout);

        final Panel fba = new Panel();
        fba.add(new Label("Fish Breed Age"));
        fishBreedScrollField = new TextField(Integer.toString(wc.getFishReprTime()), 5);
        fba.add(fishBreedScrollField);
        p.add(fba);

        final Panel sba = new Panel();
        sba.add(new Label("Shark Breed Age"));
        sharkBreedScrollField = new TextField(Integer.toString(wc.getSharkReprTime()), 5);
        sba.add(sharkBreedScrollField);
        p.add(sba);

        final Panel sst = new Panel();
        sst.add(new Label("Shark Starve Time"));
        sharkStarveScrollField = new TextField(Integer.toString(wc.getSharkEnergy()), 5);
        sst.add(sharkStarveScrollField);
        p.add(sst);
        add(p);
    }
}

File: src/main/scala/WaTor.scala
import scala.util.Random

abstract class Animal(var age: Int, val reproduceTime: Int) {
  var lastStep: Int = -1

  def duplicate: Animal
  def isFish: Boolean
  def isShark: Boolean
  private val SHARK_ENERGY_FROM_ONE_FISH = 3

  /**
   * makes one step for one element of ocean
   * @param curr - current position (column)
   * @param ocean - all world
   * @param empty - all empty columns position current position
   * @param fishes - all position with fishes around current position
   */
  def action(curr: (Int, Int), ocean: Array[Array[Animal]], empty: List[(Int, Int)], fishes: List[(Int, Int)]): Unit = {
    this match {
      case shark: Shark =>
        shark.lastStep += 1
        if(shark.energy == 0){
          ocean(curr._1)(curr._2) = null
        } else {
          if(fishes.nonEmpty || empty.nonEmpty){
            if(shark.age == shark.reproduceTime){
              shark.age = 0
              ocean(curr._1)(curr._2) = shark.duplicate
            } else {
              shark.age += 1
              ocean(curr._1)(curr._2) = null
            }
            if(fishes.nonEmpty){
              shark.energy += (SHARK_ENERGY_FROM_ONE_FISH - 1)
              val head: (Int, Int) = Random.shuffle(fishes).head
              ocean(head._1)(head._2) = shark
            } else {
              shark.energy -= 1
              val head: (Int, Int) = Random.shuffle(empty).head
              ocean(head._1)(head._2) = shark
            }
          } else {
            shark.energy -= 1
            if(shark.age == shark.reproduceTime){
              shark.age = 0
            } else {
              shark.age += 1
            }
          }
        }
      case fish: Fish =>
        fish.lastStep += 1
        if(empty.nonEmpty){
          if(fish.age == fish.reproduceTime){
            fish.age = 0
            ocean(curr._1)(curr._2) = fish.duplicate
          } else {
            fish.age += 1
            ocean(curr._1)(curr._2) = null
          }
          val head: (Int, Int) = Random.shuffle(empty).head
          ocean(head._1)(head._2) = fish
        } else {
          if(fish.age == fish.reproduceTime){
            fish.age = 0
          } else {
            fish.age += 1
          }
        }
    }
  }
}

class Shark(age: Int, reproduceTime: Int, var energy: Int) extends Animal(age, reproduceTime) {
  override def duplicate: Animal = new Shark(0, reproduceTime, 1)
  override def isFish: Boolean = false
  override def isShark: Boolean = true
}

class Fish(age: Int, reproduceTime: Int) extends Animal(age, reproduceTime) {
  override def duplicate: Animal = new Fish(0, reproduceTime)
  override def isFish: Boolean = true
  override def isShark: Boolean = false
}

class World(wh: (Int, Int), fish: (Int, Int), shark: (Int, Int, Int)) {

  private val rand = scala.util.Random
  private def newShark = new Shark(0, shark._2, rand.nextInt(shark._3-1)+1)
  private def newFish = new Fish(0, fish._2)

  private val (height, width) = wh
  private val (fishes, sharks) = (fish._1, shark._1)

  val ocean: Array[Array[Animal]] = create(fishes, sharks)

  /**
   * initialize model
   * @param fishCount - count of fishes
   * @param sharkCount - count of sharks
   * @return new ocean with animals
   */
  private def create(fishCount: Int, sharkCount: Int): Array[Array[Animal]] = {
    def randomPositions(coordinates: Set[(Int, Int)]): Set[(Int, Int)] = {
      if (coordinates.size == (fishCount + sharkCount)) coordinates
      else randomPositions(coordinates + ((rand.nextInt(height), rand.nextInt(width))))
    }
    def modifyOcean(ocean: Array[Array[Animal]], positions: Set[(Int, Int)], counter: Int): Array[Array[Animal]] = {
      if (positions.isEmpty) ocean
      else if (counter < fishCount) {
        ocean(positions.head._1)(positions.head._2) = newFish
        modifyOcean(ocean, positions.tail, counter+1)
      }
      else {
        ocean(positions.head._1)(positions.head._2) = newShark
        modifyOcean(ocean, positions.tail, counter+1)
      }
    }
    modifyOcean(Array.ofDim(height, width), randomPositions(Set()), 0)
  }

  /**
   * counts all animals of some type
   * @param ocean - all world where exist some animals
   * @param f - function which returns true if this type of animal need to be counted
   */
  private def countAll(ocean: Array[Array[Animal]])(f: Animal => Boolean): Int = {
    if(ocean.isEmpty) 0
    else ocean.head.count(f) + countAll(ocean.tail)(f)
  }

  /**
   * counts all exist fishes
   */
  def countFishes: Int = countAll(ocean)(a => null != a && a.isFish)

  /**
   * counts all exist sharks
   */
  def countSharks: Int = countAll(ocean)(a => null != a && a.isShark)

  /**
   * makes one step for all elements of the ocean
   * @param step - global number of step
   */
  def forward(step: Int): Unit = {
    0.until(height).foreach(i => 0.until(width).foreach(j => if(null != ocean(i)(j) && ocean(i)(j).lastStep < step) oneStep(i, j, step)))
  }

  /**
   * makes one step for one element of matrix
   * @param i - row
   * @param j - column
   * @param step - global number of step
   */
  def oneStep(i: Int, j: Int, step: Int): Unit = {
    val possible: List[(Int, Int)] = List(
      (i-1, j-1), (i-1, j), (i-1, j+1),
      (i, j-1), (i, j+1), (i+1, j-1),
      (i+1, j), (i+1, j+1)
    ).filter(p => 0 <= p._1 && p._1 < height && 0 <= p._2 && p._2 < width)
    /*val possible: List[(Int, Int)] = List(
      (i-1, j), (i, j-1), (i, j+1), (i+1, j)
    ).filter(p => 0 <= p._1 && p._1 < height && 0 <= p._2 && p._2 < width)*/
    val empty: List[(Int, Int)] = possible.filter(p => null == ocean(p._1)(p._2))
    val fishes: List[(Int, Int)] = possible.filter(p => null != ocean(p._1)(p._2) && ocean(p._1)(p._2).isFish)
    ocean(i)(j).action((i, j), ocean, empty, fishes)
  }
}